add build (Makefile, CI, etc) and begin to grow libraries (authz, limiter, core)

fcostin opened the pull request:
This PR adds a few building blocks & begins to grow the library that the load balancer will use:

1. established basic build scripts & CI for project (Makefile, dockerised build, github CI to run the dockerised build)
2. added dependency on open source stretchr testify library for asserting things in tests
3. created placeholder `tcplb` server application so the build scripts have a binary to build, implementation deferred to upcoming PRs
4. added libraries:
  4.1 `lib/authz` - adds `ForwardingAuthorizer` interface (which upstreams a given client may forward to), and an implementation using a static in-memory data
  4.2 `lib/limiter` - adds a `ClientReserver` interface for rate limiting clients, using the metaphor of attempting to "reserve" access to some resource, and an implementation `UniformlyBoundedClientReserver` that uses coarse-grain locking.
  4.3 `lib/core` - holds basic types that are going to be used across the libraries. I had them in `authz` for a while, the first library I added, but then every other library began to depend upon authz, which isn't a decoupled dependency structure.  Core contains: `ClientID` to represent client IDs, `Upstream` to represent an upstream, and `USet` for a set of `Upstream`s. The details of how these types will work is still under construction, but it may become clearer as the remaining libraries grow (Upstream will likely get an `Addr() net.Addr` accessor). `ClientID` might end up getting turned into a concrete value type.

If you prefer to review commit-by-commit, the commits attempt to be self contained logical commits. There is a bit of churn with some types getting moved between different library packages.

fcostin committed:
add lib/authz library for client authorisation

fcostin committed:
DEP - add github.com/stretchr/testify v1.7.1

testify dependency is added to ease unit testing.

fcostin committed:
add lib/limiter library for client rate limiting

fcostin committed:
BUILD - add Makefile, CI, tcplb server skeleton

- add placeholder server application
- add Makefile with targets to test and build
- add `make containerised_build` target, using a build container
- add basic github workflow CI, to `make containerised_build`

fcostin committed:
refactor - move clientid and upstream into lib.core

codingllama commented on Makefile:
```diff
@@ -0,0 +1,14 @@
+.PHONY: all
+
+build:
+	CGO_ENABLED=0 go build -o dist/tcplb ./cmd/tcplb/...
```
nit:

```suggestion
	CGO_ENABLED=0 go build -o dist/tcplb ./cmd/tcplb
```

The `/...` means "package and children", but when we "go build" a binary we should pick only `package main`. I'm not sure this work work if ./cmd/tcplb/ had other packages nested in it (which is a bit weird if you think about it, but it happens).

codingllama commented on builder/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+COPY go.sum .
+RUN go mod download
+
+COPY cmd cmd
+COPY lib lib
+COPY Makefile .
```
I usually copy everything in this stage and let .dockerignore filter the undesirables. It's not as tight as yours, but easier to maintain.

fcostin replied:
interesting, i haven't used .dockerignore before.  i'll leave it as is for now, i am still a bit uncomfortable with the idea of trying to specify all the things that shouldnt go in the container.

later this week if i keep tripping myself up by forgetting to copy new things into the build container, i may be motivated to revisit this

codingllama commented on builder/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM golang:1.18.1-bullseye as build-env
```
I understand pinning the patch, but I'd tempted to go with 1.18 for the ease of maintenance.

codingllama commented on builder/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+RUN make all
+
+FROM gcr.io/distroless/static
```
There are a good deal more flags needed to make a non-trivial Go binary truly static. No need to change anything, this is more of an FYI.

codingllama commented on builder/builder.sh:
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+set -e
```
Suggestion:

```suggestion
set -eu
```

Always best to start like this in a brand new script. Adding `-u` is much harder to add once the script becomes non-trivial.

fcostin replied:
thanks for flagging that. i had to look up what `set -u` means:

> Treat unset variables and parameters other than the special parameters "@" and "*" as an error when performing parameter expansion.

completely agree trying to retrofit into a large gnarly important bash script could be unpleasant

codingllama replied:
`set -eu` is my default. Some people like `-o pipefail`, but I personally don't. [Google's shell style guide](https://google.github.io/styleguide/shellguide.html) and shellcheck are good resources, if you are interested.

codingllama commented on builder/builder.sh:
```diff
@@ -0,0 +1,26 @@
+
+mkdir -p "$DIST_PATH"
+
+docker container cp "$CID":/"$TCPLB_BINARY" "$DIST_PATH"/"$TCPLB_BINARY"
```
This could easily give me a binary for the wrong OS (for example, if executed on macOS). I'm not sure this is desirable.

fcostin replied:
yes, i did wonder about that.

dist could have sub directories for different platforms, go toolchain is pretty good at cross compiling, the script & dockerised build process could be tweaked to pass in desired GOOS & GOARCH. or sweep across a supported range of (GOOS, GOARCH)

another option could be not to copy anything out and just leave the binary in the container, and ship the container image not the binary.

for challenge scope i think only one target platform is "officially supported", linux amd64

fcostin replied:
i've logged an issue for this https://github.com/fcostin/tcplb/issues/9

i suggest we dont try to fix it in scope of this PR.  if you feel this is a deal breaker please unresolve this thread.

codingllama commented on builder/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+COPY lib lib
+COPY Makefile .
+
+RUN make all
```
It'll be hard to re-run tests this way, as this is baked in this particular image layer (and won't re-execute unless a preceding layer changes).

fcostin replied:
For local development, one can use `make test` or `go test <etc>` instead of using this.

What I want the answer to be is "there is no need to re-run tests unless the input (e.g. lib, cmd, Makefile) changes, as the test output is a deterministic function of the input". But this isn't strictly true, I've already written a heavyweight unit test that might pass or fail based on the whims of goroutine scheduling if the implementation is defective.

What kind of scenarios did you have in mind where we'd want to re-run the tests?

another possible improvement could be to generate test reports and extract them out of the build container as a build artefact.

fcostin replied:
i've created an issue to track this, i propose not trying to fix this now: https://github.com/fcostin/tcplb/issues/10

but quite interested to hear your thoughts, if you have time.

please re open if you feel strongly that this needs to be fixed before we proceed

codingllama replied:
I think what got me here is that I can't quite figure out the purpose for the docker image. No objections is keeping it as-is, though.

fcostin replied:
the intent was to define a fairly controlled way to build and test the application, that is independent of people's dev environments or the build environment offered by CI.

codingllama commented on lib/README.md:
```diff
@@ -0,0 +1 @@
+tcplb libraries
```
Ditch dist/README.md and lib/README.md? They're not doing much.

fcostin replied:
(these were expedient hacks to ensure directories existed for build scripts, but its cleaner not to have them)

codingllama commented on lib/authz/backend/static/static.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// Group is a value type that represents a logical group of clients.
+type Group struct {
```
My 2c: I would make Group and UGroup "simpler" types, ie, public members and ditch New and String methods.

codingllama commented on lib/authz/backend/static/static.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+// UGroup is a value type that represents a logical group of upstreams.
+type UGroup struct {
```
```suggestion
type UpstreamGroup struct {
```

so it's clearer?

codingllama commented on lib/authz/interface.go:
```diff
@@ -0,0 +1,20 @@
+// controls which clients are allowed to forward connections to which upstreams.
+//
+// Multiple goroutines may invoke methods on a ForwardingAuthorizer simultaneously.
+type ForwardingAuthorizer interface {
```
Suggestion: ditch the interface and move the contents of "backend/static" here. It's simpler, more Go-like and "authz" is a much better package name than "backend" or "static".

In Go you'd typically export a concrete type and push the interface to the package that uses it, if an interface is needed at all.

https://github.com/golang/go/wiki/CodeReviewComments#interfaces

fcostin replied:
great link, thank you.  i think interfaces are very important to decouple designs, but moving the interface definition to the consumer side rather than producer side is reasonable. different consumers may only want different subsets of the concrete functionality & not agree on a common interface anyway.

authz is a much better package name

codingllama replied:
Yep. Go interfaces are one of the more unique aspects of the language. Because we can do "adhoc" interfaces we have much more ways to decouple, which pushes APIs to prefer concrete types.

codingllama commented on lib/core/clientid.go:
```diff
@@ -0,0 +1,8 @@
+// ClientID represents the identity of an authenticated client.
+// Implementations must have value semantics and support the
+// comparison operators (==, !=).
+type ClientID interface {
```
Similarly to ForwardingAuthorizer, I'd make this a concrete type. Ditto for Upstream.

codingllama commented on lib/core/clientid.go:
```diff
@@ -0,0 +1,8 @@
+package core
```
`core` is a bit of generic package name, but I think it passes. I'd rather see client.ID or upstream.ID, but it's not worth making those packages just to hold an ID type.

fcostin replied:
agreed. it's not as bad as "utils" but a bit of a smell...

codingllama commented on lib/core/upstream.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// USet represents a set of Upstreams
+type USet map[Upstream]struct{}
```
Is this type worth having? Why not make a `[]Upstream` or Upstream maps where you need them?

fcostin replied:
I claim the type is worth having, although there isn't much evidence to support that claim yet.

The real evidence will be when adding the forwarding logic that needs to juggle various logical sets of upstreams (authorised upstreams, healthy upstreams).

I propose leaving this as-is but have renamed it for clarity to UpstreamSet.

if you strongly disagree, please unresolve this and i can push it out of core into authz package

codingllama replied:
Wrapping a slice or map with custom utility methods is definitely something I'd flag in a regular review. That said, no need to change, let's keep it and see how it fits in future PRs.

codingllama commented on lib/authz/backend/static/static.go:
```diff
@@ -0,0 +1,84 @@
+
+func (a *Authorizer) AuthorizedUpstreams(ctx context.Context, c core.ClientID) (core.USet, error) {
+	result := core.EmptyUSet()
+	_, exists := a.config.GroupsByClientID[c]
```
```suggestion
	groups, exists := a.config.GroupsByClientID[c]
```

and reuse below?

Ditto for similar map lookups.

codingllama commented on lib/authz/backend/static/static_test.go:
```diff
@@ -0,0 +1,120 @@
+	"github.com/stretchr/testify/require"
+)
+
+type DummyClientID string
```
I'd rather use concrete types here, specially for types as simple as "client ID" and "upstream". See other comments about interface usage.

codingllama commented on lib/authz/backend/static/static_test.go:
```diff
@@ -0,0 +1,120 @@
+		c                 core.ClientID
+		cfg               Config
+		expectedUpstreams core.USet
+		expectedErr       error
```
```suggestion
```

it's always nil.

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+
+type maxReservationsExceededError struct{}
+
+func (maxReservationsExceededError) Error() string { return "maximum client reservations exceeded" }
```
Suggestion: simplify?

```suggestion
var MaxReservationsExceeded = errors.New("maximum client reservations exceeded")
```

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+// reservations for clients.
+//
+// Multiple goroutines may invoke methods on a ClientReserver simultaneously.
+type ClientReserver interface {
```
Same as other comments, I'd ditch the interface and keep only the concrete type.

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	//
+	// Implementations of ClientReserver are discouraged from blocking until
+	// a reservation becomes available.
+	TryReserve(ctx context.Context, c core.ClientID) (ClientReservation, error)
```
Maybe:

```suggestion
	TryReserve(ctx context.Context, c core.ClientID) (*ClientReservation, error)
```

so you can use `nil` instead of the "zero reservation".

fcostin replied:
removed ClientReservation entirely

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+
+	// ReleaseReservation releases a reservation that was previously acquired
+	// by TryReserve.
+	ReleaseReservation(ctx context.Context, r ClientReservation) error
```
```suggestion
	ReleaseReservation(ctx context.Context, r *ClientReservation) error
```

so you can change the reservation if you need to? Although, admittedly, there isn't much to change right now.

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	"tcplb/lib/core"
+)
+
+type ClientReservation struct {
```
I'm not sure I get what this type is for. Is it just a proxy for the client ID? Should a client ID take its place in the signatures?

Reservation instances seem to be interchangeable in practice, meaning that as long as the ReleaseReservation calls happen in the correct number, it doesn't matter _which_ reservation instance gets passed. It makes for an odd API.

fcostin replied:
great comment, thank you.  maybe "reservation" isnt the cleanest abstraction to use. 

with a struct there is a place to accommodate adding stuff in future. but YAGNI for this scope of work.

i have removed this to simplify the API. it may be fractionally less odd now.

maybe taking the reservation concept too far, a client might have preferences about which upstream to connect to (one outside with view overlooking the river?). 

> it doesn't matter which reservation instance gets passed

maybe to support tcplb's commercial model, we could make reservations non-fungible :upside_down_face:

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	// go doesn't provide a
+	// generic hash function for the application programmer to lean on. We could
+	// work around that by replacing ClientID with a concrete type or adding a
+	// Hash() uint64 method to the ClientID interface.
```
You do have the client ID, so what's wrong with `sha256.Sum256` (or whatever other hash you prefer)?

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	resByClient map[core.ClientID]int64
+}
+
+func NewUniformlyBoundedClientReserver(n int64) *UniformlyBoundedClientReserver {
```
```suggestion
func NewUniformlyBoundedClientReserver(maxReservationsPerClient int64) *UniformlyBoundedClientReserver {
```

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	n := b.resByClient[c]
+	b.sanityCheck(n)
```
I'm not sure I like this, it turns a bug into a panic. Sanity checking is fine, but let's deal with it more gracefully if we can.

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+
+func (b *UniformlyBoundedClientReserver) sanityCheck(n int64) {
+	// assert invariant 0 <= n <= MaxReservationsPerClient
+	if !((0 <= n) && (n <= b.MaxReservationsPerClient)) {
```
```suggestion
	if n < 0 || n > b.MaxReservationsPerClient {
```

a bit easier to parse, IMO.

fcostin replied:
De Morgan would approve

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+		r = ClientReservation{c: c}
+	} else {
+		err = MaxReservationsExceeded
+	}
```
Guard first, so less nesting/less declarations?

```suggestion
	if n >= b.MaxReservationsPerClient {
		return ClientReservation{}, MaxReservationsExceeded
	}

	b.resByClient[c] = n + 1
	return ClientReservation{c: c}, nil
```

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	} else {
+		err = MaxReservationsExceeded
+	}
+	b.sanityCheck(n)
```
Similarly as above, you might as well remove this and build the sanity check into the guard above.

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+func (b *UniformlyBoundedClientReserver) ReleaseReservation(ctx context.Context, r ClientReservation) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	n := b.resByClient[r.c]
+	b.sanityCheck(n)
+	n = n - 1
+	b.resByClient[r.c] = n
```
```suggestion
	b.resByClient[r.c]--
```

Avoid looking up twice? Or, if you want to keep the delete below, then do it before the 2nd lookup?

fcostin replied:
I think the delete is important.  your suggestion to do either the delete or the lookup, not both, is a good one.

codingllama commented on lib/limiter/reservation.go:
```diff
@@ -0,0 +1,133 @@
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	n := b.resByClient[r.c]
+	b.sanityCheck(n)
```
If a client wrongly calls ReleaseReservation twice this turns that error into a panic. It seems better to return a non-nil error instead.

fcostin replied:
great pick up, panic if caller makes plausible mistake in using API is not a great developer experience (and might take down entire application)

codingllama commented on lib/limiter/reservation_test.go:
```diff
@@ -0,0 +1,229 @@
+		s.Client = c
+		ctx := context.Background()
+
+		source := rand.NewSource(seed)
+		random := rand.New(source)
```
source and random redeclared? Reuse the top-level ones (or remove them)?

fcostin replied:
the docs state that non default sources aren't safe for use by multiple goroutines, so i created a new one per worker.  but your excellent suggestion to remove the randomised sleep completely also solves this issue

codingllama commented on lib/limiter/reservation_test.go:
```diff
@@ -0,0 +1,229 @@
+				s.Errors += 1
+			}
+
+			time.Sleep(time.Duration(random.Int63n(1000)))
```
Maybe sleep a fixed amount (or not at all)? I'm not sure what the randomized sleep buys you that a fixed sleep wouldn't also cause.

fcostin replied:
great point. i replaced with a small fixed sleep, removed the mutex from the implementation, test very rapidly identifies data race without randomisation.

fcostin committed:
BUILD - fix Makefile nit

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

fcostin committed:
BUILD - fix builder/builder.sh to set -u

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

fcostin committed:
lib/limiter - simplify test

timothyb89 commented on lib/core/upstream.go:
```diff
@@ -0,0 +1,50 @@
+	}
+	for k, _ := range rhs {
+		result[k] = struct{}{}
+	}
```
nit:
```suggestion
	for k := range lhs {
		result[k] = struct{}{}
	}
	for k := range rhs {
		result[k] = struct{}{}
	}
```

fcostin committed:
BUILD - unpin patch version of golang

use 1.18-bullseye

slightly less reproducibility
less toil to consume minor releases

timothyb89 commented on lib/authz/backend/static/static_test.go:
```diff
@@ -0,0 +1,120 @@
+			alice:  []Group{admin},
+			bob:    []Group{beta, alpha},
+			cindy:  []Group{beta},
+			dinesh: []Group{alpha},
```
nit: this can be simplified a bit
```suggestion
			alice:  {admin},
			bob:    {beta, alpha},
			cindy:  {beta},
			dinesh: {alpha},
```

fcostin committed:
BUILD - remove placeholder files from dist & lib

fcostin committed:
lib/limiter - rewrite invariant for clarity

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

fcostin committed:
lib/limiter - refactor for clarity

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

fcostin committed:
lib/limiter - simplify MaxReservationsExceeded

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

timothyb89 reviewed (commented):
Just a couple of small nitpicks from my end.

fcostin committed:
lib/limiter - remove ClientReserver interface

fcostin committed:
lib/limiter - update TODO, complain less about Go

fcostin committed:
lib/limiter - simplify reservation API

remove ClientReservation type. YAGNI.

fcostin committed:
lib/limiter - add NoReservationExists error

also rework internal invariant checks

fcostin committed:
lib/authz - refactor to simplify package

fcostin committed:
lib/authz - simplify tests

fcostin committed:
lib/core - refactor ClientID into concrete struct

fcostin committed:
lib/core - refactor Upstream into concrete struct

fcostin committed:
lib/core - rename USet as UpstreamSet

fcostin committed:
lib/core - simplify upstream set operations
